pypose/module/dynamics.py

```diff
@@ -0,0 +1,229 @@
+from numpy import vectorize
+import torch as torch
+import torch.nn as nn
+from torch.autograd.functional import jacobian
+
+class _System(nn.Module):
+    r'''
+    A sub-class of :obj:`torch.nn.Module` to build general dynamics.
+    
+    Args:
+        time (:obj:`boolean`): Whether the system is time-varying; defaults to False, meaning time-invariant
+    '''
+    def __init__(self, time=False):
+        super().__init__()
+        self.jacargs = {'vectorize':True, 'strategy':'reverse-mode'}
+        if time:
+            self.register_buffer('t',torch.zeros(1))
+            self.register_forward_hook(self.forward_hook)
+
+    def forward_hook(self, module, inputs, outputs):
+        self.input, self.state = inputs
+        self.t.add_(1)
+
+    def forward(self, state, input):
+        r'''
+        Parameters
+        ----------
+        state : Tensor
+                The state of the dynamic system
+        input : Tensor
+                The input to the dynamic system
+
+        Returns
+        -------
+        new_state   : Tensor
+                      The state of the system at next time step
+        observation : Tensor
+                      The observation of the system at the current step
+        '''
+
+        new_state = self.state_transition(state, input)
+        observation = self.observation(state, input)
+        return new_state, observation
+
+    def state_transition(self):
+        pass
+
+    def observation(self):
+        pass
+
+    def reset(self,t=0):
+        self.t.fill_(0) 
+
+    def set_linearization_point(self, state, input):
+        self.state, self.input = state, input
+
+    @property
+    def A(self):
+        if hasattr(self, '_A'):
+            return self._A
+        else:
+            func = lambda x: self.state_transition(x, self.input)
+            return jacobian(func, self.state, **self.jacargs)
+
+    @property
+    def B(self):
+        if hasattr(self, '_B'):
+            return self._B
+        else:
+            func = lambda x: self.state_transition(self.state, x)
+            return jacobian(func, self.input, **self.jacargs)
+
+    @property
+    def C(self):
+        if hasattr(self, '_C'):
+            return self._C
+        else:
+            func = lambda x: self.observation(x, self.input)
+            return jacobian(func, self.state, **self.jacargs)
+ 
+    @property
+    def D(self):
+        if hasattr(self, '_D'):
+            return self._D
+        else:
+            func = lambda x: self.observation(self.state, x)
+            return jacobian(func, self.input, **self.jacargs)
+    
+    @property
+    def c1(self):
+        if hasattr(self,'_c1'):
+            return self._c1
+    

```

Comment:
There should be an else as well, which returns the bias term associated with the state transition function.  Same for c2.

Reply:
I'm not sure how to implement this. In the LTI system the user gives the input for c1 and c2.